kernel/src/drivers/ps2/mod.rs

```diff
@@ -3,82 +3,85 @@ pub mod io;
 
-pub static PS2: Mutex<Ps2Controller> = Mutex::new(Ps2Controller::new());
+lazy_static! {
+    pub static ref PS2: Mutex<Ps2Controller> = Mutex::new(Ps2Controller::new());
+}
+
+bitflags! {
+    pub struct Ps2ConfigFlags: u8 {
+        const PORT_INTERRUPT_1 = 1 << 0;
+        const PORT_INTERRUPT_2 = 1 << 1;
+        const SYSTEM = 1 << 2;
+        const KEYBOARD_LOCK = 1 << 4;
+        const PORT_OUTPUT_FULL_2 = 1 << 5;
+        const PORT_TRANSLATION_1 = 1 << 6;
+    }
+}
 
 /// Represents the PS2 master controller
 pub struct Ps2Controller {
     pub initialized: bool,
     pub devices: [Ps2Device; 2],
-    config: Ps2Config,
+    config: Ps2ConfigFlags,
 }
 
 impl Ps2Controller {
-    const fn new() -> Self {
+    fn new() -> Self {
         Ps2Controller {
             initialized: false,
             devices: [
-                Ps2Device {
-                    flags: 0,
-                },
-                Ps2Device {
-                    flags: DEVICE_SECOND_FLAG,
-                }],
-            config: Ps2Config {
-                data: 0,
-            },
+                Ps2Device::new(DevicePort::Keyboard),
+                Ps2Device::new(DevicePort::Mouse),
+            ],
+            config: Ps2ConfigFlags::empty(),
         }
     }
 
     /// Initializes this PS2 controller
-    pub fn initialize(&mut self) {
+    pub fn initialize(&mut self) -> Result<(), Ps2Error> {
         println!("ps2c: initializing");
 
         for device in self.devices.iter_mut() {
-            device.disable();
-            device.set_flag(DEVICE_AVAILABLE_FLAG, false);
+            device.disable()?;
+            device.state = DeviceState::Unavailable;
         }
 
         println!("ps2c: disabled devices");
 
-        flush_output();
+        flush_output()?;
 
-        self.initialize_config();
+        self.initialize_config()?;
 
-        if !self.test_controller() {
+        if !self.test_controller()? {
             println!("ps2c: controller test failed");
         }
 
         // Test the first device
-        self.devices[0].test();
+        self.devices[0].test()?;
 
         // Check if controller supports the second device
-        if !self.config.get(ControllerConfigBit::PortOutputFull2) {
-            self.devices[1].enable();
-            self.read_config();
-            self.devices[1].disable();
+        if !self.config.contains(Ps2ConfigFlags::PORT_OUTPUT_FULL_2) {
+            self.devices[1].enable()?;
+            self.read_config()?;
+            self.devices[1].disable()?;
         }
 
         // Test the second device
-        if self.config.get(ControllerConfigBit::PortOutputFull2) {
-            self.devices[1].test();
+        if self.config.contains(Ps2ConfigFlags::PORT_OUTPUT_FULL_2) {
+            self.devices[1].test()?;
         } else {
             println!("ps2c: second device unsupported");
         }
 
         let mut available_count: u8 = 0;
         for device in self.devices.iter_mut() {
```

Comment:
This could be a fold or whatever. Alternatively, since there are only 2 devices anyway... does this need a whole loop?